src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNIndexCache.java

```diff
@@ -103,10 +106,12 @@ public synchronized void rebuild() {
      *
      * @param removalNotification key, value that got evicted.
      */
-    private void onRemoval(RemovalNotification<String, KNNIndex> removalNotification) {
-        KNNIndex knnIndex = removalNotification.getValue();
+    private void onRemoval(RemovalNotification<String, KNNIndexCacheEntry> removalNotification) {
+        KNNIndexCacheEntry knnIndexCacheEntry = removalNotification.getValue();
+
+        knnIndexCacheEntry.getFileWatcherHandle().stop();
```

Comment:
Should we add try/finally block and move ```knnIndexCacheEntry.getFileWatcherHandle().stop();``` to finally block to ensure the watcher is removed as file is deleted at this point?

Reply:
Can you clarify what would be wrapped in the try{}? ```knnIndexCacheEntry.getFileWatcherHandle().stop()``` should not throw - it just removes the watcher from the set, and doesn't declare any kind of exceptions.

Reply:
I was referring to ensuring ```knnIndexCacheEntry.getFileWatcherHandle().stop()``` being called when cache invalidate happens. some thing like this

try {
        KNNIndexCacheEntry knnIndexCacheEntry = removalNotification.getValue();
               executor.execute(() -> knnIndexCacheEntry.getKnnIndex().close());
} finally {
       knnIndexCacheEntry.getFileWatcherHandle().stop();
}

```

Reply:
Is there a perceived difference in behavior here? Both should be called regardless - in the case of the PR, ```FileWatcher.stop``` is called first, and should doesn't have an expected failure mode, and then we submit a task to ```gc``` the index.

Reply:
Agree there is no failure mode at this point. Ideally we want to ensure both gc() and stop() always get called in this function. We can defer this until we use functions that throws.